### PR TITLE
Adding revision support for import export procedure

### PR DESCRIPTION
**Pull request by 1akshitha**
This PR adds the revision support for import export process. Related Issue: https://github.com/wso2/product-apim-tooling/issues/601 and https://github.com/wso2/product-apim/issues/9406

1. If the export resource is called without the revision number, working copy will be exported. 
2. If given a number, respective revision will be exported.
3. Deployment details such as deployed environments and to whether displayOnDevportal is added a new file called deploymentEnvrionments.yaml
4. When importing the api, if deploymentEnvironment.yaml file is found, api artifact will be deployed to respective gateways.
5. When importing an api, it will first update the working copy and then a revision is created from this.

Todo after this PR:

1. This PR only provide revision support to apis. We have to fix the flow for api products as well.
2. While importing, add a validation to the environments provided.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/ImportExportAPIServiceImpl.java [hunk 1]
```diff
@@ -63,19 +64,26 @@ public File exportAPI(String apiId, String name, String version, String provider
...
             String provider = ExportUtils.validateExportParams(name, version, providerName);
             apiIdentifier = new APIIdentifier(APIUtil.replaceEmailDomain(provider), name, version);
-            api = apiProvider.getAPI(apiIdentifier);
-            apiDtoToReturn = APIMappingUtil.fromAPItoDTO(api, preserveCredentials, null);
+            apiId = APIUtil.getUUIDFromIdentifier(apiIdentifier);
         } else {
             apiIdentifier = APIMappingUtil.getAPIIdentifierFromUUID(apiId);
+        }
+
+        if (revisionNum != null) {
```
Comment (wasuradananjith) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/ImportExportAPIServiceImpl.java:79]:
```suggestion
        if (StringUtils.isNotEmpty(revisionNum)) {
```

Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/ImportExportAPIServiceImpl.java:79]: Shall we change this check like the above?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 2]
```diff
@@ -1381,6 +1381,63 @@ paths:
...
+            List of API revisions are returned.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/APIRevisionList'
+        404:
+          $ref: '#/components/responses/NotFound'
+      security:
+        - OAuth2Security:
+            - apim:api_publish
```
Comment (wasuradananjith) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:1436]: Shall we add **apim:api_import_export** scope as well?

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 3]
```diff
@@ -1722,6 +1722,25 @@ Mediation updateApiSpecificMediationPolicyContent(String apiId, Mediation mediat
...
+     * Get the revision UUID from the Revision no and API UUID
+     *
+     * @param revisionNum   No of the revision
+     * @param apiUUID       API  UUID
+     * @return UUID of the revision
+     * @throws APIManagementException if failed to get the API revision uuid
+     */
+    String getAPIRevisionUUID(String revisionNum, String apiUUID) throws APIManagementException;
+
+
```
Comment (wasuradananjith) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java:1735]: Shall we remove this empty line and add a new line after 1743?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 4]
```diff
@@ -1381,6 +1381,63 @@ paths:
...
+      x-code-samples:
+        - lang: Curl
+          source: 'curl -k -H "Authorization: Bearer ae4eae22-3f65-387b-a171-d37eaa366fa8"
+            "https://127.0.0.1:9443/api/am/publisher/v1/apis/revisions?apiId=2323&version=1.0.0&provider=adminquery=deployed:true"'
```
Comment (CrowleyRajapakse) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:1440]: Sample Code needs to add '&' before query

https://127.0.0.1:9443/api/am/publisher/v1/apis/revisions?apiId=2323&version=1.0.0&provider=admin&query=deployed:true

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 5]
```diff
@@ -1722,6 +1722,25 @@ Mediation updateApiSpecificMediationPolicyContent(String apiId, Mediation mediat
...
+    /**
+     * Get the earliest revision UUID from the revision list for a given API
+     *
+     * @param apiUUID API UUID
+     * @return Earliest revision's UUID
+     * @throws APIManagementException if failed to get the revision
+     */
+    String getEarliestRevisionUUID(String apiUUID) throws APIManagementException;
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java:1743]: Let's add a new line after this?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 6]
```diff
@@ -9787,6 +9786,33 @@ public APIRevision getAPIRevision(String revisionUUID) throws APIManagementExcep
         return apiMgtDAO.getRevisionByRevisionUUID(revisionUUID);
     }
 
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:9789]: Please remove this extra new line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 7]
```diff
@@ -9787,6 +9786,33 @@ public APIRevision getAPIRevision(String revisionUUID) throws APIManagementExcep
...
+     * @param apiUUID API UUID
+     * @return Earliest revision's UUID
+     * @throws APIManagementException if failed to get the revision
+     */
+    @Override
+    public String getEarliestRevisionUUID(String apiUUID) throws APIManagementException {
+        return apiMgtDAO.getEarliestRevision(apiUUID);
+    }
+
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:9815]: Let's remove the new line here

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportConstants.java [hunk 8]
```diff
@@ -278,4 +278,12 @@
     public static final String DEPLOYMENT_DIRECTORY= File.separator + DEPLOYMENT_DIRECTORY_NAME;
     public static final String SOURCE_ZIP_DIRECTORY_NAME = "SourceArchive.zip";
     public static final String API_PARAMS_FILE_NAME = "api_params.yaml";
+
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportConstants.java:282]: Same here

**Review by Chamindu36 — APPROVED** [review]

**Review by wasuradananjith — APPROVED** [review]

**Review by npamudika — APPROVED** [review]

**Pull request merged**